app/src/main/java/com/asfoundation/wallet/promotions/usecases/VipReferralPollingUseCase.kt

```diff
@@ -0,0 +1,49 @@
+package com.asfoundation.wallet.promotions.usecases
+
+import androidx.work.ExistingWorkPolicy
+import androidx.work.WorkManager
+import com.appcoins.wallet.gamification.repository.PromotionsRepository
+import com.appcoins.wallet.gamification.repository.entity.GamificationStatus
+import com.appcoins.wallet.gamification.repository.entity.VipReferralResponse
+import com.asfoundation.wallet.base.RxSchedulers
+import com.asfoundation.wallet.entity.Wallet
+import com.asfoundation.wallet.promotions.worker.GetVipReferralWorker
+import com.asfoundation.wallet.promotions.worker.GetVipReferralWorker.Companion.getUniqueName
+import io.reactivex.Completable
+import io.reactivex.Single
+import javax.inject.Inject
+
+class VipReferralPollingUseCase @Inject constructor(
+  private val getVipReferralUseCase: GetVipReferralUseCase,
+  private val promotionsRepository: PromotionsRepository,
+  private val repository: VipReferralPollingRepository,
+  private val workManager: WorkManager,
+  private val rxSchedulers: RxSchedulers
+) {
+
+  operator fun invoke(wallet: Wallet): Single<VipReferralResponse> {
+    return getVipReferralUseCase(wallet)
+      .doOnSuccess {
+        if (it != VipReferralResponse.invalidReferral) {
+          workManager.cancelUniqueWork(getUniqueName(wallet))
+        }
+      }
+  }
+
+  fun startPolling(wallet: Wallet): Completable =
+    promotionsRepository.getGamificationStats(wallet.address, null)
+      .subscribeOn(rxSchedulers.io)
+      .doOnNext {
+        val lastStatus =
+          repository.getLastGamificationStatus().let(GamificationStatus.Companion::toEnum)
+        if (lastStatus == GamificationStatus.APPROACHING_VIP && it.gamificationStatus == GamificationStatus.VIP) {
+          workManager.enqueueUniqueWork(
+            getUniqueName(wallet),
+            ExistingWorkPolicy.KEEP,
+            GetVipReferralWorker.getWorkRequest(wallet)
+          )
+        }
+        repository.saveLastGamificationStatus(it.gamificationStatus.toString())
+      }
+      .ignoreElements()
+}
```

Comment:
I believe this should be 2 separate use cases, mainly because a UseCase should not have more than one public method (normally the invoke method) since it goes against the logic that a use case should only have one responsibility.

Even tho this invoke method its just calling another use case that does something on success.

But by using VipReferralPollingUseCase name, it may look like calling the invoke would actually start the polling as well, but in reality, the polling is separated

Reply:
Agree 👍